crates/scheduler/src/schedule/mod.rs

```diff
@@ -0,0 +1,910 @@
+use crate::schedule::PrecedenceGraphError::{
+    Deadlock, Dependency, IncorrectSystemCompletionMessage, PendingSystemIndexNotFound,
+};
+use crossbeam::channel::{Receiver, RecvError, Select};
+use daggy::petgraph::visit::{IntoNeighbors, IntoNeighborsDirected, IntoNodeIdentifiers};
+use daggy::petgraph::{visit, Incoming};
+use daggy::{Dag, NodeIndex, WouldCycle};
+use ecs::{Schedule, ScheduleError, ScheduleResult, System, SystemExecutionGuard};
+use itertools::Itertools;
+use std::fmt::{Debug, Display, Formatter};
+use thiserror::Error;
+use tracing::{debug, error};
+
+type Sys<'system> = &'system dyn System;
+type SysDag<'system> = Dag<Sys<'system>, i32>;
+
+/// An error occurred during a precedence graph schedule operation.
+#[derive(Error, Debug)]
+enum PrecedenceGraphError {
+    /// Dependency construction error.
+    #[error("dependency from `{from}` to `{to}` in graph {graph} would cause a cycle")]
+    Dependency {
+        /// The node the dependency is coming from.
+        from: String,
+        /// The node the dependency is going to.
+        to: String,
+        /// The current state of the graph.
+        graph: String,
+    },
+    /// A pending system was completed but there is no system at its index anymore.
+    #[error("a pending system was completed but there is no system at its index anymore")]
+    PendingSystemIndexNotFound(usize),
+    /// Expected the `SystemExecutionGuard` channel to be dropped to signal completion, but got message.
+    #[error("expected the `SystemExecutionGuard` channel to be dropped to signal completion, but got message {0:?}")]
+    IncorrectSystemCompletionMessage(Result<(), RecvError>),
+    /// The schedule has deadlocked due to there being no pending systems but all systems have still not run.
+    #[error("the schedule has deadlocked due to there being no pending systems but all systems have still not run")]
+    Deadlock,
+}
+
+/// Whether a precedence graph operation succeeded.
+type PrecedenceGraphResult<T, E = PrecedenceGraphError> = Result<T, E>;
+
+/// Orders [`System`]s based on their precedence.
+#[derive(Default, Clone)]
+pub struct PrecedenceGraph<'systems> {
+    dag: SysDag<'systems>,
+    /// Systems which have been given out, and are awaiting execution.
+    ///
+    /// Warning: Node indices are _not_ stable and will be invalidated if [`dag`] is mutated.
+    pending: Vec<(Receiver<()>, NodeIndex)>,
+    /// Which systems have already been executed this tick.
+    ///
+    /// Warning: Node indices are _not_ stable and will be invalidated if [`dag`] is mutated.
+    already_executed: Vec<NodeIndex>,
+}
+
+impl<'systems> Debug for PrecedenceGraph<'systems> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        // Instead of printing the struct, format the DAG in dot-format and print that.
+        // (so it can be viewed through tools like http://viz-js.com/)
+        writeln!(f, "{:?}", daggy::petgraph::dot::Dot::new(self.dag.graph()))
```

Comment:
A get alot of false positives in this file from daggy in CLion. Are you also experiencing this? Here I get:
```
`Dot<&DiGraph<Sys, i32>>` doesn't implement `Debug` (required by {:?})
```

Reply:
Yup, it's horrible... 

I searched for solutions and found a bunch of bug reports about it, but they were all marked resolved so we must be encountering a new case. 

I tried looking into ways of disabling that error for this file, but it seems that there's no way to disable errors using comments and the like, only warnings. :(